Add bit-level circuit reasoning: circuit/extens tactics and bind command

Gustavo2622 opened the pull request:
Introduce a bit-blasting backend that discharges bitvector/Hoare goals by
translating them to boolean circuits and checking them with an SMT solver,
together with the surface commands and tactics that drive it.

Library:
  - libs/lospecs: hash-consed AIG circuit representation (aig), bitvector/
    register operations (circuit), dependency analysis (deps), and an SMT
    decision procedure backed by Bitwuzla (smt); adds the bitwuzla-cxx
    dependency.

Binding (theories/datatypes/QFABV.ec):
  - `bind bitstring|array|op|circuit ...` maps EasyCrypt types and operators
    onto their bit-level models (new CRB_* parse nodes); side conditions are
    left as `realize` goals.

Tactics:
  - `circuit`: closes first-order, Hoare, and equivalence goals by checking
    the corresponding circuit valid/equivalent; `circuit simplify` rewrites
    a goal using the circuit normal form.
  - `proc change circuit ... + w { s }`: replaces a program fragment with a
    circuit-equivalent one (new `Change variant of the program-rewrite node).
  - `extens [v] : t`: extensionality over list-all and finite bitstring
    enumeration.

Supporting infrastructure:
  - ecLowCircuits / ecCircuits: form <-> circuit translation with lazy
    integer substitution and alpha-invariant memoisation (ecAlphaInvHashtbl);
    ecPhlBDep wires the tactics into the PHL goal types; ecTypesafeFol helpers.

Docs and tests:
  - doc/tactics/{circuit,bindings,extens}.rst.
  - tests/circuits/: unit tests extracted from the documentation examples.

Gustavo2622 committed:
Circuit reasoning functionality

strub commented on src/ecCircuits.ml:
```diff
@@ -0,0 +1,1114 @@
+  | `Record
+  | `Hoare
+  | `Instr
+] 
```
Isn't one exception whose argument is a sum-type would be better?

strub committed:
LSP

strub committed:
LSP

strub committed:
LSP

Gustavo2622 committed:
First pass on new error reporting

Gustavo2622 committed:
Added implicit type translation

Gustavo2622 committed:
Added duplicate checking for spec files and general cleanup

Gustavo2622 committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

Gustavo2622 committed:
Killed warnings

Gustavo2622 committed:
Error and printing improvements

Gustavo2622 committed:
Added fail case circuit tests

Gustavo2622 committed:
Minor error propagation fix

Gustavo2622 committed:
Allow uninitialized program variable in circuit solve for hoare goals

Gustavo2622 committed:
Changed alpha equiv hash to not compute form normal form and fixed circuit test

Gustavo2622 committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

Gustavo2622 committed:
Fixed merge problems + nits on error printing

Gustavo2622 committed:
Fixing FIXMEs

Gustavo2622 committed:
Moved map reference to inside function scope in hash

Gustavo2622 commented on src/ecCircuits.ml:
```diff
@@ -0,0 +1,1176 @@
+        Why3.Hashcons.combine (doit st tp) i 
+      | FhoareF _hF -> 
+          assert false
+(*      FIXME: do we want this case and the one below?
```
Do we want to handle this?

Gustavo2622 commented on src/ecCircuits.ml:
```diff
@@ -0,0 +1,1176 @@
+    | Fpvar   (pv, mem) -> 
+      let v = match pv with
+      | PVloc v -> v
+      (* FIXME: Should globals be supported? *)
```
Do we want to support globals for this merge?

Gustavo2622 commented on src/ecCircuits.ml:
```diff
@@ -0,0 +1,1176 @@
+    | FequivS  _
+    | FeagerF  _
+    | Fpr _ -> circ_error (CantConvertToCirc `Hoare) 
+    (* FIXME: do we want to allow conversion of hoare statements? 
```
Do we want to be able to generate a circuit from these things?

Gustavo2622 commented on src/ecCircuits.ml:
```diff
@@ -0,0 +1,1176 @@
+      false
+    end
+
+(* FIXME: add support for spec bindings for abstract/opaque operators 
```
Not used currently, but do we want this for the merge? Should not be too hard to do

Gustavo2622 commented on src/ecCircuits.ml:
```diff
@@ -0,0 +1,1176 @@
+
+  let st1 = close_circ_lambda st1 in 
+  let st2 = close_circ_lambda st2 in
+  (* FIXME: what was the intended behaviour for keep? *)
```
FIXME

Gustavo2622 commented on src/ecCircuits.ml:
```diff
@@ -0,0 +1,1176 @@
+        EcEnv.notify env `Debug "Failed to translate hypothesis for var %s with error %a, skipping@." (tostring_internal id) (pp_circ_error ppe) e;
+        try 
+          open_circ_lambda st [(id, ctype_of_ty env t)]
+        (* FIXME PR: Should only catch circuit translation errors, hack *)
```
Fixed, remove the comment

Gustavo2622 committed:
Started documenting circuit tactic and small fix to circuit test

Gustavo2622 committed:
Added documentation and tests for circuit tactics

strub committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

strub committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

Gustavo2622 committed:
Remove bindings for hidden theories

Gustavo2622 committed:
WIP: Fixing memory leaks

Gustavo2622 committed:
Fixed alpha-invariant hashing bug

Gustavo2622 committed:
Revert spec file in easycrypt.project and extend loader to support spec files

Gustavo2622 committed:
Improve locate API

Gustavo2622 committed:
Proper error message

strub committed:
load spec file relatively to current processed file

Gustavo2622 committed:
Error message for ec scope

Gustavo2622 committed:
nits

Gustavo2622 committed:
WIP: Bitwuzla -> Bitwuzla_cxx

strub committed:
progress on clones & bindings

strub committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

strub committed:
generic bindings now work

strub committed:
Merge remote-tracking branch 'origin/bdep_ecCircuitsRefactor' into bdep_ecCircuitsRefactor

strub committed:
[refold]: allow rigid unification

Syntax: `rewrite -/~(pattern)`

strub committed:
QFABV: aligned extraction

Gustavo2622 committed:
WIP: Proc change add code + simple framing

strub committed:
Fix oppath cloning

strub committed:
Fix QFABV

Gustavo2622 committed:
Fixes to proc change indexes

Gustavo2622 committed:
Revert "Fixes to proc change indexes"

This reverts commit 1790498ca125ccc699c9ba1d8ada77e751ac00b2.

Gustavo2622 committed:
Revert "WIP: Proc change add code + simple framing"

This reverts commit d78990265a5276aa86238e6b5bd8ee1984b45a22.

strub committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

strub committed:
fix merge warnings

strub committed:
int -> int64 for lospec literals

strub committed:
lospec: raw equality

strub committed:
Forward call with framed pre

strub committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

strub committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

strub committed:
Merge remote-tracking branch 'origin/vscode' into bdep_ecCircuitsRefactor

strub committed:
Tighten proc-change observability

Restrict proc-change local postconditions to variables observable from the
continuation or the goal post, instead of equating all writes from the
replaced fragment.

strub committed:
Merge branch 'proc-change-no-eq-for-post' into bdep_ecCircuitsRefactor

strub committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

strub committed:
Merge branch 'vscode' into bdep_ecCircuitsRefactor

strub committed:
Merge branch 'vscode' into bdep_ecCircuitsRefactor

strub committed:
fix dune

strub committed:
Merge remote-tracking branch 'origin/main' into bdep_ecCircuitsRefactor

Gustavo2622 committed:
Fixed circuit test binding syntax

strub committed:
Fix simplify flag handling in `cfold`

strub committed:
Merge branch 'fix-cfold-simplify' into bdep_ecCircuitsRefactor

strub committed:
Merge branch 'bdep_ecCircuitsRefactor' of github.com:EasyCrypt/easycrypt into bdep_ecCircuitsRefactor

Gustavo2622 committed:
Added precondition attachment to non-equality goals for bdep

strub committed:
Add goal printing flags (-upto, -lastgoals) and LLM agent guide

Add two new flags for the `easycrypt` CLI to support LLM coding agents:
- `-upto <pos>`: compile up to a given position and print goals there
- `-lastgoals`: print the last unproven goals

Also add a dedicated `llm` command mode and an LLM agent guide
(doc/llm/CLAUDE.md) documenting EasyCrypt tactics and workflow for
use with AI coding assistants.

strub committed:
Replace LLM batch mode with interactive REPL protocol

Replace the batch `easycrypt llm file.ec` command with an interactive
REPL (`easycrypt llm`) using a line-oriented protocol over stdin/stdout.

The REPL supports:
- LOAD "file.ec" [LINE[:COL]] — compile a file, optionally up to a line
- UNDO / REVERT <uuid|name> — navigate proof state
- GOALS / GOALS ALL — inspect current or all subgoals
- CHECKPOINT <name> — named bookmarks for branching exploration
- SEARCH <pattern> — lemma search
- QUIET ON/OFF — suppress goal display for bulk tactic application
- Direct EasyCrypt input (tactics, declarations, search, print, etc.)

Responses use a typed envelope (OK/ERROR with uuid) and <END> sentinel
for reliable parsing by LLM agent frameworks.

Additional improvements:
- LOAD reports the last processed line in the response tag
- Error messages include the source tactic text
- Only the current subgoal is shown by default (with remaining count)
- LOAD uses loc_end for upto check (more intuitive line semantics)

Update doc/llm/CLAUDE.md with the new protocol reference, interactive
workflow guide, and proof strategy tips from practical usage.

strub committed:
Add LLM REPL improvements and built-in guide

New REPL features:
- LOAD -nosmt: skip SMT calls during prefix compilation (uses
  Proofs:weak pragma), restores strict mode for interactive tactics
- Error responses now include the current goal state, eliminating
  the need for a separate GOALS roundtrip after failures
- Multi-line input support via <BEGIN>/<DONE> delimiters
- HELP meta-command and -help flag: print the LLM agent guide
  (doc/llm/CLAUDE.md) from within the REPL or at startup

The guide is installed as part of the distribution (doc site) so it
stays version-matched with the binary. Agent frameworks can use
`easycrypt llm -help` to populate their system prompt automatically.

Documentation updates:
- REVERT-to-LOAD-uuid workflow (instant restart vs recompile)
- Pattern-based search syntax with examples
- SMT usage guidance (pure arithmetic/logic only)
- rewrite-in-H clarification

strub committed:
Merge branch 'llm-interactive' into bdep_ecCircuitsRefactor

# Conflicts:
#	dune
#	src/ecEnv.ml
#	src/ecHiTacticals.ml
#	src/ecOptions.ml
#	src/ecPV.ml
#	src/ecPV.mli
#	src/phl/ecPhlExists.ml
#	src/phl/ecPhlRewrite.ml
#	tests/forward-call.ec

Gustavo2622 committed:
Fixed proc change not registering new vars in program memory

Gustavo2622 committed:
Add bit-level circuit reasoning: circuit/extens tactics and bind command

Introduce a bit-blasting backend that discharges bitvector/Hoare goals by
translating them to boolean circuits and checking them with an SMT solver,
together with the surface commands and tactics that drive it.

Library:
  - libs/lospecs: hash-consed AIG circuit representation (aig), bitvector/
    register operations (circuit), dependency analysis (deps), and an SMT
    decision procedure backed by Bitwuzla (smt); adds the bitwuzla-cxx
    dependency.

Binding (theories/datatypes/QFABV.ec):
  - `bind bitstring|array|op|circuit ...` maps EasyCrypt types and operators
    onto their bit-level models (new CRB_* parse nodes); side conditions are
    left as `realize` goals.

Tactics:
  - `circuit`: closes first-order, Hoare, and equivalence goals by checking
    the corresponding circuit valid/equivalent; `circuit simplify` rewrites
    a goal using the circuit normal form.
  - `proc change circuit ... + w { s }`: replaces a program fragment with a
    circuit-equivalent one (new `Change variant of the program-rewrite node).
  - `extens [v] : t`: extensionality over list-all and finite bitstring
    enumeration.

Supporting infrastructure:
  - ecLowCircuits / ecCircuits: form <-> circuit translation with lazy
    integer substitution and alpha-invariant memoisation (ecAlphaInvHashtbl);
    ecPhlBDep wires the tactics into the PHL goal types; ecTypesafeFol helpers.

Docs and tests:
  - doc/tactics/{circuit,bindings,extens}.rst.
  - tests/circuits/: unit tests extracted from the documentation examples.

Co-authored-by: Pierre-Yves Strub <pierre-yves.strub@pqshield.com>
